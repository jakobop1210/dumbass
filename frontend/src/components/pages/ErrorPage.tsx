import Button from '../atoms/Button'
import Breadcrumb from '../atoms/Breadcrumb'

/**
 * NotFoundPage component - A user-friendly error page for routing failures
 *
 * This component is displayed when the user tries to navigate to a route that does not exist.
 * It provides an error message and a button to navigate back to the homepage.
 */

export type ErrorPageType = {
    title: string
    subTitle: string
    description: string
    buttonText: string
    buttonFunction: () => void
}

export const ErrorPage = (props: ErrorPageType) => {
    return (
        <main>
            <Breadcrumb
                items={[
                    {
                        name: 'Unknown Page'
                    }
                ]}
            />
            <section className='flex items-center pt-10 px-5 sm:pt-32 sm:px-10 flex-col text-gray-500 text-center'>
                <header>
                    <h1 className='text-7xl'>{props.title}</h1>
                    <h2 className='text-3xl my-4'>{props.subTitle}</h2>
                </header>

                <section className='max-w-xl mb-10'>
                    <p>{props.description}</p>
                </section>
                <section className='w-1/6 min-w-fit'>
                    <Button
                        title={props.buttonText}
                        type='button'
                        onClick={() => props.buttonFunction()}
                    />
                </section>
            </section>
        </main>
    )
}
