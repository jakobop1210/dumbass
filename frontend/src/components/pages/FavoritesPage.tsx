import { useSelector } from 'react-redux'
import { RootState } from '../../redux/store'
import { Artist, Song } from '../../lib/types'
import { ClipLoader } from 'react-spinners'
import Breadcrumb from '../atoms/Breadcrumb'
import CardView from '../organisms/CardView'
import { ArtistCardProps, SongCardProps } from '../molecules/ArtistSongCard'
import { useFavorites } from '../../hooks/useFavorites'

export default function FavoritesPage() {
    const username = useSelector((state: RootState) => state.user.username)

    const { dataSongs, loading, error, dataArtists } = useFavorites(username)

    // Processing and mapping data for Card components
    const artists: Artist[] = dataArtists?.getArtistsByIds || []
    const songs: Song[] = dataSongs?.getSongsByIds || []

    const artistCardData = artists.map(
        (artist: Artist): ArtistCardProps => ({
            cardType: 'artist',
            id: artist.id,
            title: artist.name,
            alternateNames: artist.alternate_names,
            imageUrl: artist.image_url,
            rating: artist.average_rating,
            numOfRatings: artist.number_of_ratings
        })
    )

    const songCardData = songs.map(
        (song: Song): SongCardProps => ({
            cardType: 'song',
            id: song.id,
            title: song.title,
            artist: song.artist_names,
            imageUrl: song.header_image_url,
            rating: song.average_rating,
            numOfRatings: song.number_of_ratings,
            releaseDate: song.release_date
        })
    )

    if (!username) {
        return (
            <div className='h-96 flex items-center justify-center text-blueGray italic'>
                <h1>You need to be logged in to add favorites...</h1>
            </div>
        )
    }

    return (
        <main className='w-full flex flex-col'>
            <Breadcrumb items={[{ name: 'Favorites' }]} />
            <section className='w-full flex flex-col justify-center items-center mt-5'>
                {loading ? (
                    <div className='h-96 flex items-center'>
                        <ClipLoader color={'#8fc0a9'} size={100} />
                    </div>
                ) : error ? (
                    <div className='h-96 flex items-center'>
                        <h1 className='text-blueGray text-2xl'>
                            Error loading data :/
                        </h1>
                    </div>
                ) : (
                    <>
                        <section className='w-full flex justify-center'>
                            <CardView
                                title='Favorited Songs'
                                cardData={songCardData}
                                customErrorMessage='You have no favorite songs...'
                            />
                        </section>
                        <section className='w-full flex justify-center'>
                            <CardView
                                title='Favorited Artists'
                                cardData={artistCardData}
                                customErrorMessage='You have no favorite artists...'
                            />
                        </section>
                    </>
                )}
            </section>
        </main>
    )
}
