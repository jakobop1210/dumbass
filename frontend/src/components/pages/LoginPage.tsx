import { FaMusic } from 'react-icons/fa'
import { useState } from 'react'
import { useNavigate } from 'react-router-dom'
import InputField from '../atoms/InputField'
import Button from '../atoms/Button'
import { customToast } from '../../lib/utils'
import { LOGIN_USER } from '../../graphql/mutations/userMutations'
import { useMutation } from '@apollo/client'
import { useDispatch } from 'react-redux';
import { setUserLogin, setUserName } from '../../redux/actions/userActions';

/**
 * LoginPage component - Used for user authentication
 *
 * Allows a user to log in by providing a username and a password. After successful login, navigation to the homepage is performed and a successful login message is shown. If the login fails, an error message is displayed.
 *
 * @param {LoginPageProps} props - Properties passed down from parent component. Optionally includes `setUser`.
 */
export default function LoginPage(): JSX.Element {
    const navigate = useNavigate()
    const dispatch = useDispatch()
    const [username, setUsername] = useState('')
    const [password, setPassword] = useState('')
    const [loginUserMutation] = useMutation(LOGIN_USER)

    /**
     * loginUser - Event handler for form submission.
     *
     * Validates provided username and password. If credentials match (username: 'guest', password: 'guest'), navigates to the homepage and optionally invokes the `setUser` function if provided.
     * Shows a toast notification for successful login or failure.
     *
     * @param {React.FormEvent} e - Event object related to the form submission.
     */
    async function loginUser(e: React.FormEvent) {
        e.preventDefault()

        try {
            const result = await loginUserMutation({
                variables: { username, password }
            })
            const { data } = result

            if (data.loginUser) {
                customToast('success', 'Successfully logged in')
                dispatch(setUserLogin()); // Dispatch Redux action to set user as logged in
                dispatch(setUserName(username)); // Dispatch Redux action to set username
               
                // Store username and login status in local storage
                localStorage.setItem('username', username);
                localStorage.setItem('isLoggedIn', 'true');

                navigate('/')
            } else {
                customToast('error', 'An error occurred while logging in')
            }
        } catch (error) {
            console.log(JSON.stringify(error, null, 2))
            customToast('error', 'Wrong password or username')
        }
    }

    return (
        <main className='w-screen flex justify-center'>
            <section className='flex h-4/5 sm:w-[32rem] w-4/5 flex-col justify-center px-8 py-10 my-10 rounded-lg md:px-20 bg-white'>
                <header className='sm:mx-auto sm:w-full sm:max-w-sm'>
                    <div className='flex items-center justify-center'>
                        <FaMusic size={60} color='#8FC0A9' />
                    </div>
                    <h2 className='mt-10 text-center text-3xl font-bold leading-9 tracking-tight text-blueGray'>
                        Sign in to your account
                    </h2>
                </header>

                <div className='mt-10 sm:mx-auto sm:w-full sm:max-w-sm'>
                    <form className='space-y-6' onSubmit={loginUser}>
                        <InputField
                            id='username'
                            type='username'
                            title='Username'
                            value={username}
                            onChange={setUsername}
                            required
                        />
                        <InputField
                            id='password'
                            type='password'
                            title='Password'
                            value={password}
                            onChange={setPassword}
                            required
                        />
                        <Button title={'Sign in'} type={'submit'} />
                    </form>

                    <footer className='mt-10 text-center text-sm text-blueGray'>
                        Don't have an account?
                        <button
                            onClick={() => navigate('/register')}
                            className='ml-2 font-semibold leading-6 text-green hover:cursor-pointer'>
                            Register here
                        </button>
                    </footer>
                </div>
            </section>
        </main>
    )
}
