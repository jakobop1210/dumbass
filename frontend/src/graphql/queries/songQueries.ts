import { gql } from '@apollo/client'

/**
 * GraphQL query: GET_SONG_BY_ID.
 *
 * This query fetches detailed information about a song by its unique ID. 
 * The details include the song's lyrics, header image URL, release date, title, artist names, average rating, and the number of ratings.
 *
 * @param {string} $id - The unique ID of the song to fetch.
 * @returns {Object} An object containing detailed information about the song.
 */
export const GET_SONG_BY_ID = gql`
    query ($id: ID!) {
        getSongById(id: $id) {
            lyrics
            header_image_url
            release_date
            title
            artist_names
            average_rating
            number_of_ratings
        }
    }
`

/**
 * GraphQL query: GET_SONGS_BY_ID.
 *
 * This query retrieves detailed information for a list of songs based on their unique IDs.
 * The details include the song's ID, title, artist names, header image URL, release date, primary artist ID, average rating, and number of ratings.
 *
 * @param {Array<number>} $ids - An array of IDs of the songs to fetch.
 * @returns {Array<Object>} An array of objects, each containing detailed information about a song.
 */
export const GET_SONGS_BY_ID = gql`
    query ($ids: [Int]!) {
        getSongsByIds(ids: $ids) {
            id
            title
            artist_names
            header_image_url
            release_date
            primary_artist_id
            average_rating
            number_of_ratings
        }
    }
`

/**
 * GraphQL query: GET_SONGS_ON_TITLE.
 *
 * This query fetches a list of songs based on a search term, with optional parameters for pagination and sorting.
 * The details returned for each song include the header image URL, ID, title, artist names, average rating, number of ratings, and release date.
 *
 * @param {string} [$title] - The title or part of the title of the songs to search for.
 * @param {number} [$limit] - The maximum number of songs to return.
 * @param {string} [$sort] - The sorting criteria (e.g., 'rating', 'alphabetical').
 * @param {number} [$page] - The page number for pagination.
 * @returns {Array<Object>} An array of objects, each containing information about a song.
 */
export const GET_SONGS_ON_TITLE = gql`
    query ($title: String, $limit: Int, $sort: String, $page: Int) {
        getSongsOnTitle(
            title: $title
            limit: $limit
            sort: $sort
            page: $page
        ) {
            header_image_url
            id
            title
            artist_names
            average_rating
            number_of_ratings
            release_date
        }
    }
`

/**
 * GraphQL query: COUNT_SONGS.
 *
 * This query counts the number of songs in the database that match a given title or search term.
 *
 * @param {string} [$title] - The title or part of the title of the songs to count.
 * @returns {number} The count of songs matching the search criteria.
 */
export const COUNT_SONGS = gql`
    query ($title: String) {
        countSongs(title: $title)
    }
`
