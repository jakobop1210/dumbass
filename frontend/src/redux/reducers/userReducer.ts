import { createReducer } from '@reduxjs/toolkit';
import { setUserLogin, setUserLogout, setUserName } from '../actions/userActions';

interface UserState {
    loggedIn: boolean;
    username: string | null;
}

const initialState: UserState = {
    loggedIn: false,
    username: null,
};

const userReducer = createReducer(initialState, (builder) => {
    builder
        .addCase(setUserLogin, (state) => {
            state.loggedIn = true;
        })
        .addCase(setUserLogout, (state) => {
            state.loggedIn = false;
            state.username = null; // Nullify username
        })
        .addCase(setUserName, (state, action) => {
            state.username = action.payload; // Set username
        });
});

export default userReducer;
