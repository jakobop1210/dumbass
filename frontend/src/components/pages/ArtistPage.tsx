import { useNavigate, useParams } from 'react-router-dom'
import { InfoPageTemplate } from '../organisms/InfoPageTemplate'
import { useQuery } from '@apollo/client'
import { GET_ARTIST_BY_ID } from '../../graphql/queries/artistQueries'
import { faCircleInfo } from '@fortawesome/free-solid-svg-icons'
import { formatAlternateNames } from '../../lib/utils'
import { ErrorPage } from './ErrorPage'

/**
 * `ArtistPage` Component.
 *
 * This component fetches and displays detailed information about an artist using an `InfoPageTemplate`.
 * It queries the data for a specific artist by their ID using GraphQL and renders the `InfoPageTemplate` with this data.
 * The page displays the artist's name, alternate names, image, rating, and description. It includes a tab for additional artist information.
 *
 * If the data is not available or an error occurs, the component renders an `ErrorPage` component,
 * providing the user with a message and an option to navigate back to the homepage.
 *
 * @returns {JSX.Element} The rendered page with either the artist information or an error message.
 */
export const ArtistPage = () => {
    const { id = '' } = useParams()
    const navigate = useNavigate()

    // GraphQL query to fetch artist details by ID
    const { data, loading } = useQuery(GET_ARTIST_BY_ID, {
        variables: { id: parseInt(id) }
    })
    
    // Conditional rendering based on data availability or loading status
    if (data?.getArtistById || loading) {
        return (
            <InfoPageTemplate
                isLoading={loading}
                title={data?.getArtistById.name}
                subtitle={formatAlternateNames(
                    data?.getArtistById.alternate_names
                )}
                image={data?.getArtistById.image_url}
                averageRating={data?.getArtistById.average_rating}
                numOfRatings={data?.getArtistById.number_of_ratings}
                description={data?.getArtistById.description}
                tabs={[
                    {
                        title: 'Info',
                        icon: faCircleInfo
                    }
                ]}
                id={id}
                type='artist'
            />
        )
    } else {
        return (
            <ErrorPage
                title='Oops!'
                subTitle='Unknown error'
                description='Something wrong happened, you can try again or click the button below to go back to the homepage'
                buttonText='Go to homepage'
                buttonFunction={() => navigate('/')}
            />
        )
    }
}
