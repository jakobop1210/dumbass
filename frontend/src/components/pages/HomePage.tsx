import { useState } from 'react'
import { ArtistCardProps, SongCardProps } from '../molecules/ArtistSongCard'
import CardView from '../organisms/CardView'
import CommonSearchBar from '../molecules/CommonSearchBar'
import { GET_TOP_ARTISTS } from '../../graphql/queries/artistQueries'
import { GET_TOP_SONGS } from '../../graphql/queries/songQueries'
import { useQuery } from '@apollo/client'
import { customToast } from '../../lib/utils'
import { ClipLoader } from 'react-spinners'

// Define a type for an artist
type Artist = {
    id: string
    name: string
    image_url: string
    alternate_names: string[]
    average_rating: number
    number_of_ratings: number
}

// Define a type for a song
type Song = {
    id: string
    title: string
    artist_names: string
    header_image_url: string
    release_date: string
    primary_artist_id: number
    average_rating: number
    number_of_ratings: number
}

/**
 * @component ExplorePage
 *
 * `ExplorePage` is a React functional component that displays a page where users
 * can explore various artists and songs using a search bar and view them via card layouts.
 */
export default function HomePage() {
    const [filter, setFilter] = useState('Song')
    const limitData = 12

    // Define GraphQL queries using the useQuery hook
    const {
        loading: loadingArtists,
        error: errorArtists,
        data: dataArtists
    } = useQuery(GET_TOP_ARTISTS, {
        variables: { limit: limitData }
    })
    const {
        loading: loadingSongs,
        error: errorSongs,
        data: dataSongs
    } = useQuery(GET_TOP_SONGS, {
        variables: { limit: limitData }
    })

    if (errorArtists || errorSongs) {
        console.log(
            `Error: ${JSON.stringify(errorArtists || errorSongs, null, 2)}`
        )

        customToast('error', 'Error', 'Could not load data')
    }

    // Combine and display data
    const artists: Artist[] = dataArtists?.getTopArtists || []
    const songs: Song[] = dataSongs?.getTopSongs || []

    const artistCardData: ArtistCardProps[] = artists.map((artist: Artist) => ({
        cardType: 'artist',
        id: artist.id,
        title: artist.name,
        alternateNames: artist.alternate_names,
        imageUrl: artist.image_url,
        rating: artist.average_rating,
        numOfRatings: artist.number_of_ratings
    }))

    const songCardData: SongCardProps[] = songs.map((song: Song) => ({
        cardType: 'song',
        id: song.id,
        title: song.title,
        artist: song.artist_names,
        imageUrl: song.header_image_url,
        rating: song.average_rating,
        numOfRatings: song.number_of_ratings,
        releaseDate: song.release_date
    }))

    return (
        <main className='flex flex-col items-center justify-center w-screen'>
            <CommonSearchBar
                className='w-4/5 mt-10 drop-shadow mb-10'
                filterOptions={['Song', 'Artist']}
                selectedFilter={filter}
                onFilterChange={(newFilter) => setFilter(newFilter)}
            />
            {loadingArtists || loadingSongs ? (
                <div className='h-96 flex items-center'>
                    <ClipLoader color={'#8fc0a9'} size={100} />
                </div>
            ) : (
                <>
                    {errorArtists || errorSongs ? (
                        <div className='h-96 flex items-center'>
                            <h1 className='text-blueGray text-2xl'>
                                Error loading data :/
                            </h1>
                        </div>
                    ) : (
                        <>
                            <section className='w-full flex justify-center'>
                                <CardView
                                    title='Top Songs'
                                    cardData={songCardData}
                                />
                            </section>
                            <section className='w-full flex justify-center'>
                                <CardView
                                    title='Top Artists'
                                    cardData={artistCardData}
                                />
                            </section>
                        </>
                    )}
                </>
            )}
        </main>
    )
}
