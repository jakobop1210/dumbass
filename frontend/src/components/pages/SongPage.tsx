import { useNavigate, useParams } from 'react-router-dom'
import { InfoPageTemplate } from '../organisms/InfoPageTemplate'
import { useQuery } from '@apollo/client'
import { GET_SONG_BY_ID } from '../../graphql/queries/songQueries'
import { faMicrophoneLines } from '@fortawesome/free-solid-svg-icons'
import { ErrorPage } from './ErrorPage'

/**
 * `SongPage` Component.
 *
 * This component is responsible for displaying detailed information about a specific song.
 * It fetches the song's data using a GraphQL query based on the song's ID obtained from the URL parameters.
 * The component renders an `InfoPageTemplate` with the fetched song data, including its title, artist names, image, rating, lyrics, and release date.
 * 
 * The `InfoPageTemplate` is used to structure and present the information in a consistent and informative layout.
 * The page includes a tab for lyrics, represented by a microphone icon, to indicate the lyrical content of the song.
 *
 * If the data for the song is not available or an error occurs during fetching, the `ErrorPage` component is rendered,
 * displaying an error message and providing an option to navigate back to the homepage.
 *
 * @returns {JSX.Element} The rendered page for the song with detailed information or an error message.
 */
export const SongPage = () => {
    const { id = '' } = useParams()
    const navigate = useNavigate()

    const { data, loading } = useQuery(GET_SONG_BY_ID, {
        variables: { id: parseInt(id) }
    })

    if (data?.getSongById || loading) {
        return (
            <InfoPageTemplate
                isLoading={loading}
                title={data?.getSongById.title}
                subtitle={data?.getSongById.artist_names}
                image={data?.getSongById.header_image_url}
                averageRating={data?.getSongById.average_rating}
                numOfRatings={data?.getSongById.number_of_ratings}
                lyrics={data?.getSongById.lyrics}
                release_date={data?.getSongById.release_date}
                tabs={[
                    {
                        title: 'Lyrics',
                        icon: faMicrophoneLines
                    }
                ]}
                id={id}
                type='song'
            />
        )
    } else {
        return (
            <ErrorPage
                title='Oops!'
                subTitle='Unknown error'
                description='Something wrong happened, you can try again or click the button below to go back to the homepage'
                buttonText='Go to homepage'
                buttonFunction={() => navigate('/')}
            />
        )
    }
}
