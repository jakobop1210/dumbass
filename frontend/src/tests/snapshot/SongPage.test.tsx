import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { vi } from 'vitest';
import { configureStore } from '@reduxjs/toolkit';
import userReducer from '../../redux/reducers/userReducer';
import { Provider } from 'react-redux';
import { GET_SONG_BY_ID } from '../../graphql/queries/songQueries';
import { SongPage } from '../../components/pages/SongPage';

vi.mock('react-router-dom', () => ({
    useParams: () => ({ id: '1' }),
    useNavigate: () => vi.fn(),
    useLocation: () => ({ pathname: '/song/1', state: {} }),
    // eslint-disable-next-line
    BrowserRouter: require('react-router-dom').BrowserRouter,
}));

const mockStore = configureStore({
    reducer: {
        user: userReducer,
    },
    preloadedState: {
        user: { loggedIn: true, username: 'testUser' },
    },
});

// Mock for GraphQL-spÃ¸rring
const mocks = [
    {
        request: {
            query: GET_SONG_BY_ID,
            variables: { id: 1 },
        },
        result: {
            data: {
                getSongById: {
                    title: 'Test Song',
                    artist_names: ['Test Artist'],
                    header_image_url: 'http://example.com/song.jpg',
                    average_rating: 4.2,
                    number_of_ratings: 80,
                    lyrics: 'Test Lyrics',
                    release_date: '2022-01-01',
                },
            },
        },
    },
];

test('SongPage snapshot', () => {
    const { asFragment } = render(
        <Provider store={mockStore}>
            <MockedProvider mocks={mocks} addTypename={false}>
                <Router>
                    <SongPage />
                </Router>
            </MockedProvider>
        </Provider>
        
    );
    expect(asFragment()).toMatchSnapshot();
});
