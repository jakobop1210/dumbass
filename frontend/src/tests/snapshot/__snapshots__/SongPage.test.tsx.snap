// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SongPage snapshot 1`] = `
<DocumentFragment>
  <main>
    <div
      class="w-1/2 text-left absolute"
    >
      <nav
        aria-label="breadcrumb"
        class="w-max"
      >
        <div
          class="py-2 px-4"
          role="status"
        >
          <span
            aria-busy="true"
            aria-live="polite"
          >
            <span
              class="react-loading-skeleton"
              style="width: 100px; height: 20px;"
            >
              ‌
            </span>
            <br />
          </span>
        </div>
      </nav>
    </div>
    <section
      class="flex items-center justify-center w-screen sm:p-12 lg:py-16 lg:px-32"
    >
      <article
        class="min-[900px]:grid min-[900px]:grid-cols-4 w-full max-w-4xl gap-10 bg-white sm:rounded-xl shadow p-5 pt-10 xs:p-10"
      >
        <header
          class="flex flex-col xs:flex-row gap-5 justify-start xs:items-center md:block md:col-span-1 xs:mb-5 md:m-0"
        >
          <figure
            class="aspect-w-1 aspect-h-1 md:w-full"
          >
            <span
              aria-busy="true"
              aria-live="polite"
            >
              <span
                class="react-loading-skeleton"
                style="width: 100%; height: 175px;"
              >
                ‌
              </span>
              <br />
            </span>
          </figure>
          <section
            class="md:mt-5 flex flex-col gap-2 xs:gap-2"
          >
            <span
              aria-busy="true"
              aria-live="polite"
            >
              <span
                class="react-loading-skeleton"
                style="width: 75%; height: 20px;"
              >
                ‌
              </span>
              <br />
            </span>
            <span
              aria-busy="true"
              aria-live="polite"
            >
              <span
                class="react-loading-skeleton"
                style="width: 50%; height: 15px;"
              >
                ‌
              </span>
              <br />
            </span>
            <span
              aria-busy="true"
              aria-live="polite"
            >
              <span
                class="react-loading-skeleton"
                style="width: 75%; height: 25px;"
              >
                ‌
              </span>
              <br />
            </span>
            <span
              aria-busy="true"
              aria-live="polite"
            >
              <span
                class="react-loading-skeleton"
                style="width: 65%; height: 20px;"
              >
                ‌
              </span>
              <br />
            </span>
            <span
              aria-busy="true"
              aria-live="polite"
            >
              <span
                class="react-loading-skeleton"
                style="width: 100%; height: 40px;"
              >
                ‌
              </span>
              <br />
            </span>
          </section>
        </header>
        <section
          class="col-span-3"
        >
          <nav
            class="border-b border-gray-200 dark:border-gray-700"
          >
            <ul
              class="flex flex-wrap -mb-px text-sm font-medium text-center text-gray-500 dark:text-gray-400"
            >
              <ul
                class="flex gap-3 justify-around"
              >
                <span
                  aria-busy="true"
                  aria-live="polite"
                >
                  <span
                    class="react-loading-skeleton"
                    style="width: 100px; height: 50px;"
                  >
                    ‌
                  </span>
                  <br />
                </span>
                <span
                  aria-busy="true"
                  aria-live="polite"
                >
                  <span
                    class="react-loading-skeleton"
                    style="width: 100px; height: 50px;"
                  >
                    ‌
                  </span>
                  <br />
                </span>
              </ul>
            </ul>
          </nav>
          <article
            class="flex flex-col gap-8 py-8 xs:px-8"
          />
        </section>
      </article>
    </section>
  </main>
</DocumentFragment>
`;
